### ISSUE TITLE
Alpha channel bit depth

### ISSUE THREAD

**Issue by robin-raymond** [kornelski/pngquant#340]
This is more of a question. If I needed a palette output restricted to ARGB4888 and not ARGB8888 is it possible somehow to modify the code to reduce the bit depth available? I see the code that handles the IE 6.0 and alpha and it made me think that if I were to modify the alpha input values to only contains values that could be reduced to 4 bits (i.e. fixed intervals with no alpha values between those intervals) would that cause the output to strictly stick to those alphas in the output palette or would it generate blended alpha values beyond the restricted alphas I allowed in order to optimize the palette? I know I can go in and force the output palette to only have the alpha values I want but then the palette could end up with wasteful intra-alpha values that get reduced to the same alpha value in the end.

**Comment by kornelski** [comment]
You can preprocess the input, which will help significantly, but is not a guarantee that output will be equally posterized (since colors are blended).

A better clue would be the `posterize` feature. Currently it assumes all channels are posterized equally, but you could make it per channel.

**Comment by robin-raymond** [comment]
@kornelski thanks! I'm going to look into that.

**Comment by robin-raymond** [comment]
````
        px.r = posterize_channel(px.r, posterize);
        px.g = posterize_channel(px.g, posterize);
        px.b = posterize_channel(px.b, posterize);
        px.a = posterize_channel(px.a, posterize);
````

Is it really as simple as passing a unique value per channel here? if so, then this is dirt simple to change.

**Comment by kornelski** [comment]
Yup. There's also posterization mask in pam.c hash generation. That's some ugly bit-fiddling there.

**Comment by robin-raymond** [comment]
I see what you mean in that routine! Greatly appreciated. Follow up question: Let's say I needed to add a new non-standard channel to the palette, so instead of ARGB4888 it was DARGB448888 where "D" stands for depth. Is there any reason why modifying the code everywhere to handle a new D "fake color" and treating the depth similar to a color wouldn't work? I know it would be painful since argb is everywhere but it at least seems pretty straight forward where the color routines are used throughout the code (although I may have to change some alloc assumptions of course).

**Comment by kornelski** [comment]
Color channels are gamma-corrected, and use premultiplied alpha (so when alpha = 0, *all* channels are 0), so you'd have to avoid correction and premultiplication (or perhaps scale your depth accordingly?)

SIMD routine for color comparison assumes 4-float layout.

**Comment by robin-raymond** [comment]
Right. Good points. Should be fun :) Thanks again! That answers all my questions.
